### PR TITLE
Fix "workspaceComplete" fragment selection matching

### PR DESCRIPTION
**Pull request by ThatParticularPencil**
**Summary**
Solves #3484.
Makes `SearchAgent.workspaceComplete` more robust when the LLM returns description-shaped fragment selections instead of copying `fragmentid` values exactly from the Workspace TOC. Also adds tests for fragment matching.

`app/src/main/java/ai/brokk/agents/SearchAgent.java`
- Clarifies the `workspaceComplete` tool parameter description so the LLM is told to copy the `fragmentid` attribute verbatim, with a UUID-shaped example.
- Preserves the existing strict matching order:
  1. exact fragment ID
  2. exact fragment description
  3. tolerant description fallback
- Adds a more tolerant fallback for description-shaped selections:
  - trims input
  - lowercases with `Locale.ROOT` (new import)
  - collapses repeated whitespace
  - accepts normalized prefix matches only when they resolve to exactly one fragment
- Reports ambiguous tolerant matches as invalid selections with candidate fragment IDs, so the next retry can use a UUID instead of guessing.

`app/src/test/java/ai/brokk/agents/SearchAgentTest.java`
- Adds coverage for exact fragment ID and description matching.
- Adds coverage for normalized/prefix description matching against a real SearchAgent TOC shape.
- Adds coverage for ambiguous normalized/prefix matches.
- Adds coverage confirming retry recovery already preserves accepted IDs: the first call can include one accepted ID plus one invalid selection, and the second call only needs to send the corrected selection.

**Notes**
I did not actually recreate the issue. This PR covers the reported failure mode directly with tests: an LLM returning description-derived strings rather than the UUID-shaped `fragmentid` value from the TOC.

The tolerant matching is intentionally conservative. Loose matching only runs for selections that would previously have been rejected, and it refuses to guess when multiple fragment descriptions match. Some smarter heuristics could be added here, or more aggressive normalization. However, because the instructions now indicate a preference for fragment IDs, those changes are likely unnecessary.

### CODE REVIEW

File: app/src/main/java/ai/brokk/agents/SearchAgent.java [hunk 1]
```diff
@@ -581,12 +583,39 @@ private FragmentSelectionResolution resolveFragmentSelections(List<String> suppl
                 continue;
             }
 
+            var looseMatchingIds = resolveLooseDescriptionSelection(token, descriptionToIds);
+            if (looseMatchingIds.size() == 1) {
```
Comment (foundev) [inline app/src/main/java/ai/brokk/agents/SearchAgent.java:587]: **Edge case: normalized-exact match should win over normalized-prefix match.**

Consider a workspace with two descriptions: `"User"` and `"User class"`. If the LLM sends `"user"` (lowercase paraphrase of the first), the strict path at L577 misses (case-sensitive). Then `resolveLooseDescriptionSelection` runs:

- `"user".startsWith("user")` → true → matches `"User"`
- `"user class".startsWith("user")` → true → matches `"User class"`

…and gets rejected as ambiguous, even though `"user"` is a normalized-exact match for exactly one description.

Suggested fix: in the loose path, prefer entries where `normalizedDescription.equals(normalizedToken)` over prefix matches. If any normalized-equal entries exist, return only those; otherwise fall back to bidirectional `startsWith`. Equivalently, you could insert a fourth tier between L584 and L586:

```java
var normalizedExactIds = resolveNormalizedExactSelection(token, descriptionToIds);
if (normalizedExactIds.size() == 1) {
    accepted.add(normalizedExactIds.getFirst());
    continue;
}
```

This would also let you tighten the prefix path's contract — it only ever runs for *non-equal* fuzzy matches.

File: app/src/main/java/ai/brokk/agents/SearchAgent.java [hunk 2]
```diff
@@ -581,12 +583,39 @@ private FragmentSelectionResolution resolveFragmentSelections(List<String> suppl
...
     }
 
+    private static List<String> resolveLooseDescriptionSelection(
+            String token, Map<String, List<String>> descriptionToIds) {
+        var normalizedToken = normalizeFragmentSelection(token);
+        return descriptionToIds.entrySet().stream()
+                .filter(entry -> {
+                    var normalizedDescription = normalizeFragmentSelection(entry.getKey());
+                    return normalizedDescription.startsWith(normalizedToken)
+                            || normalizedToken.startsWith(normalizedDescription);
```
Comment (foundev) [inline app/src/main/java/ai/brokk/agents/SearchAgent.java:608]: Two related concerns on this filter:

**1. No minimum-length guard.** A very short token (`"S"`, `"a"`) can resolve to a single fragment whenever the workspace has just one description starting with that letter. That's not a paraphrase of the description; it's noise being accepted. Worth requiring `normalizedToken.length() >= 3` (or comparable) before treating prefix matches as authoritative — the strict ID/description paths still cover legitimate short identifiers.

**2. The second branch (`normalizedToken.startsWith(normalizedDescription)`) is untested.** All four new tests exercise the LLM-shortens-the-description direction. If this branch is meant to absorb cases like the LLM appending `" implementation"` or class-type info to a description, please add a test; if not, drop it — keeping an untested asymmetric branch in fuzzy-match code is the kind of thing that comes back as a bug later.

The `.distinct()` at L611 is also dead code in the current shape: each fragment ID lives in exactly one map entry, so `flatMap` never produces duplicates. Harmless, but worth removing if it's not load-bearing for some future shape of `descriptionToIds`.

File: app/src/main/java/ai/brokk/agents/SearchAgent.java [hunk 3]
```diff
@@ -581,12 +583,39 @@ private FragmentSelectionResolution resolveFragmentSelections(List<String> suppl
                 continue;
             }
 
+            var looseMatchingIds = resolveLooseDescriptionSelection(token, descriptionToIds);
+            if (looseMatchingIds.size() == 1) {
+                accepted.add(looseMatchingIds.getFirst());
+                continue;
+            } else if (looseMatchingIds.size() > 1) {
+                bad.add(token + " (ambiguous description, matches multiple IDs: " + looseMatchingIds + ")");
```
Comment (foundev) [inline app/src/main/java/ai/brokk/agents/SearchAgent.java:591]: This message is byte-identical to L582, but the two scenarios are meaningfully different from the LLM's perspective:

- **L582:** the LLM sent an *exact* description and multiple fragments genuinely share that description → only a UUID can disambiguate.
- **L591:** the LLM sent a paraphrase that happens to prefix multiple descriptions → it could either send a UUID *or* send a more specific description.

Distinguishing them gives the model better recovery options. Something like:

```java
bad.add(token + " (matches multiple fragments via fuzzy match; provide a UUID or a more specific description: " + looseMatchingIds + ")");
```

Minor, but the whole retry mechanism only works if the model can act on the message text, so wording carries weight.

File: app/src/test/java/ai/brokk/agents/SearchAgentTest.java [hunk 4]
```diff
@@ -91,6 +91,154 @@ void workspaceComplete_retryGuidanceOmitsFurtherInvestigation() throws Interrupt
...
+        registry = new ToolRegistry().builder().register(agent).build();
+
+        var directive = agent.buildTurnDirective(1, List.of());
+        assertTrue(directive.contains("description=\"Summary of ai.brokk.acp.AcpProtocol class\""));
+        assertTrue(directive.contains("fragmentid=\"" + protocolId + "\""));
+
+        var result = executeTool(
+                "workspaceComplete",
+                """
+                {"fragmentIdsOrDescriptions":["summary   of AI.BROKK.ACP.ACPPROTOCOL"]}
```
Comment (foundev) [inline app/src/test/java/ai/brokk/agents/SearchAgentTest.java:156]: This test (and `rejectsAmbiguousNormalizedDescriptionPrefix` below) only covers the LLM-shortens-the-description direction. The companion direction in `resolveLooseDescriptionSelection` — `normalizedToken.startsWith(normalizedDescription)`, where the LLM emits *more* than the description — has no test.

If the second branch is intentional, add a case like:

```java
{"fragmentIdsOrDescriptions":["Summary of ai.brokk.acp.AcpProtocol class implementation details"]}
```

…against a fragment whose description is exactly `"Summary of ai.brokk.acp.AcpProtocol class"`, and assert it resolves to that fragment. If the branch isn't intentional (see code comment at L608), this test isn't needed.

**Review by foundev — COMMENTED** [review]
Solid PR — addresses every point in #3484 and the new tests are well-targeted. A few suggestions inline; none are blocking, but the first one (normalized-exact priority) is a real edge case that the current logic mishandles. The rest are quality/coverage improvements.

Not in scope for review, but worth flagging: this fixes the deterministic failure modes of `resolveFragmentSelections`. The original symptom in #3484 described a *broken session*, which suggests the LLM kept sending description-shaped strings even after the retry message. The stronger `@P` text should help, but there's no end-to-end test against a flaky model — that's a separate problem and probably not worth solving here.
